### PR TITLE
Fix issue 123: PReLU with non-shared parameters

### PR DESCRIPTION
**Pull request by ilya-edrenkin**
Fixing issue #123 
The code now works identically to CPU version of nn.PReLU on 1D and 2D tensors.

I think it would be a good idea to provide a clear specification what this unit shall do with 3D and 4D tensors (for the CPU version as well) or just restrict its usage to 1D and 2D, but this seems to be outside of the scope of this fix.

### CODE REVIEW

**Comment by szagoruyko** [comment]
have you checked if it works with 3d and 4d? seems to me that this will break it
we need better tests to cover all cases

**Comment by ilya-edrenkin** [comment]
I would be really glad to write such a test, however I'm not sure what is the expected behavior for 3D and 4D. Even for the CPU version (untouched by this commit) I get quite unexpected results for 3D and 4D:
```lua
th> prelu = nn.PReLU(3)
                                                                      [0.0001s]
th> prelu.weight = torch.Tensor({2, 4, 6})
                                                                      [0.0001s]
th> prelu:forward(torch.ones(2, 3):mul(-1))
-2 -4 -6
-2 -4 -6
[torch.DoubleTensor of size 2x3]

                                                                      [0.0041s]
th> prelu:forward(torch.ones(2, 1, 3):mul(-1))
/home/ilia/torch/install/share/lua/5.1/nn/PReLU.lua:14: wrong number of input planes at /home/ilia/torch/extra/nn/generic/PReLU.c:48
stack traceback:
	[C]: in function 'PReLU_updateOutput'
	/home/ilia/torch/install/share/lua/5.1/nn/PReLU.lua:14: in function 'forward'
	[string "_RESULT={prelu:forward(torch.ones(2, 1, 3):mu..."]:1: in main chunk
	[C]: in function 'xpcall'
	/home/ilia/torch/install/share/lua/5.1/trepl/init.lua:630: in function 'repl'
	...ilia/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:185: in main chunk
	[C]: at 0x00406240
                                                                      [0.0002s]
th> prelu:forward(torch.ones(2, 1, 1, 3):mul(-1))
/home/ilia/torch/install/share/lua/5.1/nn/PReLU.lua:14: wrong number of input planes at /home/ilia/torch/extra/nn/generic/PReLU.c:48
stack traceback:
	[C]: in function 'PReLU_updateOutput'
	/home/ilia/torch/install/share/lua/5.1/nn/PReLU.lua:14: in function 'forward'
	[string "_RESULT={prelu:forward(torch.ones(2, 1, 1, 3)..."]:1: in main chunk
	[C]: in function 'xpcall'
	/home/ilia/torch/install/share/lua/5.1/trepl/init.lua:630: in function 'repl'
	...ilia/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:185: in main chunk
	[C]: at 0x00406240
```

I do understand that the reason is this check in PReLU.c:
```c
      if(input->size[(input_ndim + 1) % 2] != nOutputPlane)
	THError("wrong number of input planes");
```
However I failed to understand how this is supposed to work. Probably I missed some part of documentation on usage of PReLU?

**Comment by szagoruyko** [comment]
yes this is expected. For 3D your input has to be 3,H,W, for 4D - B,3,H,W if you give nOutputPlane = 3 as a parameter. It is applied per map.

**Comment by ilya-edrenkin** [comment]
OK, got it, thanks. I will rewrite it to support this behavior.
(I think it would not harm to reflect this convention somewhere in the documentation of nn.PReLU, since it is probably not completely obvious).

**Comment by ilya-edrenkin** [comment]
I've updated the code so that now it works identical to CPU version on 1D, 2D, 3D and 4D tensors (I've checked both outputs and gradients, including weights' gradients).

**Comment by szagoruyko** [comment]
can be merged I think.

**Comment by soumith** [comment]
hey @ilya-edrenkin could you squash your commits? looks good to go.

**Comment by ilya-edrenkin** [comment]
sure. done.

**Comment by soumith** [comment]
thanks a lot mate!

**Comment by ilya-edrenkin** [comment]
thanks for the awesome torch project!
